### ISSUE TITLE
windows build broken due to missing `install`

### ISSUE THREAD

**Issue by newville** [tomopy/tomopy#371]
Trying to build from current master with Anaconda Python 3.6 on Windows (64 bit), gives

C:\Users\xas_user\Documents\GitHub\tomopy>python setup.py build
running build
running build_py
running egg_info
writing tomopy.egg-info\PKG-INFO
writing dependency_links to tomopy.egg-info\dependency_links.txt
writing top-level names to tomopy.egg-info\top_level.txt
reading manifest file 'tomopy.egg-info\SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files matching '__pycache__' found under directory '*'
warning: no previously-included files matching '*.pyc' found under directory '*'
warning: no previously-included files matching '*.o' found under directory 'config'
writing manifest file 'tomopy.egg-info\SOURCES.txt'
copying tomopy\sharedlibs\libtomopy.dll -> build\lib.win-amd64-3.6\tomopy\sharedlibs
running build_ext
INFO:__main__:Config for Windows
INFO:__main__:COMPILER_DIR is C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\MinGW\bin\gcc.exe
INFO:__main__:Config output:
## generated by build.py  Tue Dec  4 09:24:58 2018
COMPILER_DIR   = C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin
SHAREDLIB      = libtomopy.dll
ARCH_TARGET    =
LINK_LIB       = -LC:/Users/xas_user/AppData/Local/Continuum/Anaconda3 -LC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/Libra
ry/bin -LC:/Users/xas_user/AppData/Local -LC:/Windows/System32
INCLUDE        = -IC:/Users/xas_user/Documents/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include
 -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/Library/include
CONDA_COMPAT   =
INSTALL_PREFIX = C:/Users/xas_user/Documents/GitHub/tomopy/build/lib.win-amd64-3.6
####

C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o art.o ../src/art.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o bart.o ../src/bart.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o fbp.o ../src/fbp.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o grad.o ../src/grad.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o gridrec.o ../src/gridrec.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o mlem.o ../src/mlem.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o morph.o ../src/morph.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o osem.o ../src/osem.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o ospml_hybrid.o ../src/ospml_hybrid.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o ospml_quad.o ../src/ospml_quad.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o pml_hybrid.o ../src/pml_hybrid.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o pml_quad.o ../src/pml_quad.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o prep.o ../src/prep.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o project.o ../src/project.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o remove_ring.o ../src/remove_ring.c
../src/gridrec.c: In function 'gridrec':
../src/gridrec.c:397:39: warning: 'filphase_iter' may be used uninitialized in this function [-Wmaybe-uninitialized]
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o sirt.o ../src/sirt.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o stripe.o ../src/stripe.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o tv.o ../src/tv.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o utils.o ../src/utils.c
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -mdll -O -DMS_WIN64 -DUSE_MKL -Wall -IC:/Users/xas_user/Docu
ments/GitHub/tomopy/include -IC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/include -IC:/Users/xas_user/AppData/Local/Contin
uum/Anaconda3/Library/include  -DPY3K -DWIN32 -std=c99    -c -o vector.o ../src/vector.c
../src/utils.c: In function 'calc_coords':
../src/utils.c:132:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:137:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c: In function 'calc_dist':
../src/utils.c:276:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:282:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:288:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:308:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:313:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:318:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:323:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:328:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:333:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:338:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:343:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:348:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c: In function 'calc_dist2':
../src/utils.c:361:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
../src/utils.c:369:0: warning: ignoring #pragma omp simd [-Wunknown-pragmas]
C:/Users/xas_user/AppData/Local/Continuum/Anaconda3/MinGW/bin/gcc.exe -shared -s -LC:/Users/xas_user/AppData/Local/Continuum/Anaco
nda3 -LC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/Library/bin -LC:/Users/xas_user/AppData/Local -LC:/Windows/System32/Lib
rary -LC:/Users/xas_user/AppData/Local/Continuum/Anaconda3 -LC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/Library/bin -LC:/
Users/xas_user/AppData/Local -LC:/Windows/System32/Library/bin -LC:/Users/xas_user/AppData/Local/Continuum/Anaconda3 -LC:/Users/xa
s_user/AppData/Local/Continuum/Anaconda3/Library/bin -LC:/Users/xas_user/AppData/Local -LC:/Windows/System32/libs -LC:/Users/xas_u
ser/AppData/Local/Continuum/Anaconda3 -LC:/Users/xas_user/AppData/Local/Continuum/Anaconda3/Library/bin -LC:/Users/xas_user/AppDat
a/Local -LC:/Windows/System32/PCBuild/amd64 -lvcruntime140 -o ../tomopy/sharedlibs/libtomopy.dll art.o bart.o fbp.o grad.o gridrec
.o mlem.o morph.o osem.o  ospml_hybrid.o ospml_quad.o pml_hybrid.o pml_quad.o prep.o project.o  remove_ring.o sirt.o stripe.o tv.o
 utils.o vector.o -lm -lmkl_rt
install -m 755 ../tomopy/sharedlibs/libtomopy.dll C:/Users/xas_user/Documents/GitHub/tomopy/build/lib.win-amd64-3.6/tomopy/sharedl
ibs/
make: install: Command not found
Mk.base:31: recipe for target `install' failed
make: *** [install] Error 127
Traceback (most recent call last):
  File "build.py", line 186, in <module>
    build_libtomopy()
  File "build.py", line 60, in build_libtomopy
    subprocess.check_call(tuple(cmd))
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '('make', '-j4', '-f', 'Makefile.windows', 'install')' returned non-zero exit status 2.
Traceback (most recent call last):
  File "setup.py", line 67, in <module>
    'Programming Language :: C']
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\site-packages\setuptools\__init__.py", line 143, in setup
    return distutils.core.setup(**attrs)
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\distutils\core.py", line 148, in setup
    dist.run_commands()
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\distutils\dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\distutils\dist.py", line 974, in run_command
    cmd_obj.run()
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\distutils\command\build.py", line 135, in run
    self.run_command(cmd_name)
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\distutils\cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\distutils\dist.py", line 974, in run_command
    cmd_obj.run()
  File "setup.py", line 18, in run
    self.build_extension(ext)
  File "setup.py", line 24, in build_extension
    '--prefix={}'.format(extdir)))
  File "C:\Users\xas_user\AppData\Local\Continuum\Anaconda3\lib\subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '('C:\\Users\\xas_user\\AppData\\Local\\Continuum\\Anaconda3\\python.exe', 'build.py', '--p
refix=C:\\Users\\xas_user\\Documents\\GitHub\\tomopy\\build\\lib.win-amd64-3.6')' returned non-zero exit status 1.

This is because there is no `install` program installed as part of the anaconda build environment on Windows.   The conda packages for Make, MinGW, and setuptools are all installed.   I don't see an `install` included in any of the standalone MinGW distributions either, and do not understand how the appveyor environment works for this. 
 
This was broken by @jrmadsen as part of #355.  In that discussion @jrmadsen made many questionable assertions about why the build system "needed to be changed" but *clearly* promised they knew what they were doing and would not break the build on Windows.   

Please fix soon.

**Comment by jrmadsen** [comment]
Somewhere in one of the MinGW directories is the `install` command. I can't fire up my Parallels version right now bc I need to renew the license but I tested it on Windows. Go to the directory above where the `make` command is installed, do a search for it, and then at that to the PATH and it will work

**Comment by jrmadsen** [comment]
It is not broken. This is essentially a bug-report claiming you don't have `python` bc python was installed to `/opt/local/bin/python` and you only have `PATH=/usr/bin:/usr/local/bin`. The `install` command is the recommended way of doing installations in Makefiles.

**Comment by jrmadsen** [comment]
`conda install -c msys2 m2-coreutils` specifically has `install.exe`. Although compilers are large packages -- you should just install a system compiler and put it in the system PATH. I keep mine in `C:\MinGW` and add these to the PATH: `C:\MinGW\bin`, `C:\MinGW\msys\1.0\bin`, and `C:\MinGW\mingw32\bin`

**Comment by carterbox** [comment]
I was **not** able to reproduce this problem on my Windows test machine, and as mentioned, the Appveyor build passes. 

@newville As a reminder, `conda install` is our [recommended installation method](https://tomopy.readthedocs.io/en/latest/install.html). If you are building the commits between releases, you should expect to have some difficulty. 

Thanks to @jrmadsen for some suggestions to start solving this problem.

**Comment by jrmadsen** [comment]
AppVeyor is working bc it has system compilers installed outside of conda and those external installations are providing the `install.exe`. The MinGW package you get from conda is not the full package -- it is stripped down to conserve size.

**Comment by newville** [comment]
@jrmadsen @carterbox  Err, no it most definitely *is* broken, and this is not a bug report about having the path incorrectly set to find a particular file.  I encourage you to strive for civility and deference to people reporting bugs and be less dismissive.   I am trying to help you.  It appears you are recognizing this.

The conda environment to build tomopy does include MinGW and Make, which should make it fully capable to build libtomopy and tomopy, without any external build tools installed.  To be clear,  this *did* work and now does not (hence the term "broken").   Indeed, such an environment does *build* libtomopy, but does not install it, because the `install` utility is not installed within the conda environment.  

Of course, if you had `install` somewhere else in your path, it would probably work.  But `install` is not made available by the conda environment defined for tomopy.   That is the problem here.

It might be feasible to say that something like `msys2` or `coreutils` could be a requirement for building `tomopy`, but it currently is not.   I would say that adding such a dependency just for the simple use of `install` here is not warranted when "copy" ('cp' on posix, 'copy' on windows) would do just fine for the use case here.  That is, I think the main benefit of `install` vs `cp` (carefully erasing then creating a new file in place of any existing file to allow apps running with the older lib to continue working) is not relevant here ("installation" is to an intermediate location to be finally copied by setuptools) and would not work on Windows anyway.   Having separate "copy" and "change mode" commands should not be a very big deal.

At the very least, set `INSTALL=copy` in `Makefile.windows` and `INSTALL=install -m 775` on linux and darwin, and then use `${iNSTALL}` in `Mk.base`.  Basically, if you have a command hardwired in your Makefile, someone somewhere will run into a problem.  That `rm -f` in the `clean` directive looks pretty fragile to me. 

You should probably not be using an external MinGW for the appveyor build but rather use the MinGW conda package -- that does get installed anyway, which is sort of puzzling.

@carterbox I do understand your recommended install method.  I am trying to build conda packages for another channel.  The code should build, and did build, and would build except that the build process got broken.  It shouldn't be hard or controversial to fix.

**Comment by jrmadsen** [comment]
If you have a solution, why haven't you put in a pull request? I'm hearing quite a bit of "y'all need to do X and Y" and "X looks fragile" but it would help if there was some actual collaborative effort. I split my time with other projects and I can only spend so much time on this sort of stuff, esp when it comes to Windows. I actually _wanted_ to just do a copy from the start! The docs can be updated now that we have your issue on the books but, at the same time, you are using a GNU compiler and build system that has been ported to Windows. They are not native Windows tools, issues like this are very common and the solution is easy: you install a base toolchain at the OS level as a fallback to provide general tools (who seriously knows the difference between `make-v14.3` and `make-v18.5`?) and go from there. Why would you want to duplicate the same packages, especially large packages like compiler, in every single environment when the versions are not critical and the interface to the command line utils that they provide doesn't change? That's just a waste of hard-drive space IMO.

From your perspective, it might not matter which compiler is used and you may not see a difference between the Anaconda compilers and the other ones on your system, but I can definitively state that it does matter quite a bit. For example, I am seeing significant speedup on the supercomputers (almost 2x for some problems) when I use the Intel compilers and I have only been able to analyze this performance and test out the optimizations bc of these updates I have provided. 

So, in summation, I am trying to maintain your Windows build but there are other factors at play here and be forewarned, there are more changes to come and I will do my best to mitigate them but there is only so much I can reasonably do.

**Comment by jrmadsen** [comment]
Quick side note to the base toolchain -- the version of a package/library is most important when you are _linking_ it to another library bc the symbols in the library change between versions. But for tools like `install`, `make`, `cmake`, etc. they are just generating a product. Any changes to the tool between version are internal to the package and there no dependency so the only real difference between the versions are the features. Thus, unless you need newer features from more recent releases, having a base toolchain is completely safe and portable because nothing from those tools is getting injected into the product that could cause compatibility issues later.

**Comment by dgursoy** [comment]
I'll use this thread as an opportunity to ask few questions that are somehow relevant. 

@jrmadsen @carterbox @newville: Have you ever tried the Windows Subsystem for Linux? It is a new Windows 10 feature that enables you to run native Linux command-line tools directly on Windows, alongside your traditional Windows desktop and modern store apps.

More information here:
https://docs.microsoft.com/en-us/windows/wsl

I am not trying to say that we should use this, but since it has the potential to reduce some of the OS support overhead of TomoPy in the future, I think it is worth investigating. Right now I am only curious how (also if) this works and see if anyone tried it.

**Comment by carterbox** [comment]
@newville I'm trying to help you too. I've tried to recreate your problem, but I was unsuccessful. I'm sorry that your alternate build was broken, but our build tests don't include your build environment. I'm offering help make your build environment match the one we test on, so I need more information about what and how you are trying to build. If you want to take a code editing approach, then please open a pull.

>But install is not made available by the conda environment defined for tomopy. That is the problem here.

I don't believe this is true. Are you using the Anaconda prompt? `install.exe` is included with all of my Windows Anaconda installations. It's located at `minconda3\Library\usr\bin\install.exe`. If you are using the Anaconda prompt. Please send a copy of your conda build recipe.

>You should probably not be using an external MinGW for the appveyor build but rather use the MinGW conda package -- that does get installed anyway, which is sort of puzzling.

That's my fault. However, it is irrelevant to this problem. I tested using the MinGW conda package and the Appveyor build still works.

**Comment by jrmadsen** [comment]
@dgursoy I have used the WSL and it works very well. I explored using it quite a bit when it first was released as a beta bc another project I work on had a pthreads implementation so multithreading wasn't supported on Windows and the OpenGL + Windows graphics were minimally supported. With the WSL and a Xming server running in Windows, I was able to get both working. But in the advent of Docker, it seems to just amount to a (persistent) built-in Docker container without the flexibility of changing the container's base OS. You can create persistent Docker containers with volumes so, in the end, I don't think it provides any extra benefits. 

Along those lines, maybe tomopy should register as an organization and create a Docker Hub repository, e.g. https://hub.docker.com/u/tomopy/. Aside from it making it very easy to use run tomopy without having to configure/build anything, it would add these benefits:

- using TomoPy + Intel compilers when the user doesn't have access to an Intel license to the compilers: e.g. `tomopy/release:intel`
- containers with TomoPy already installed (obviously) `tomopy/release:latest`
- containers with the entire toolchain setup but tomopy not installed: `tomopy/devel:latest`
- containers with different features enabled: `tomopy/release:cuda10.0-ubuntu18.04`, `tomopy/devel:cuda9.2-ubuntu18.04`, etc.

With that in place, I think tomopy could _consider_ the possibility of directly supporting Windows since it wouldn't restrict Windows users at all for developing and/or using tomopy. I have no idea how many tomopy + Windows users there actually are so that idea in said in the context of my own ignorance. In any event, if @newville is open to the idea, tomopy could do this and explicitly break the Windows builds in a release (e.g. intentionally removing a couple `#ifdef` for Windows) and if hell is raised on the issues board bc there are actually a lot of Windows users, a quick patch could be reapplied and the idea could be shelved.

**Comment by newville** [comment]
@carterbox I have two different Windows machines that show the same behavior -- there is *not* a file named `install.exe` under Anaconda3, including in any `pkgs` folders.  Both machines are using Anaconda3 5.2.0, and are using Python 3.6.7.  I have done 

```
conda install mkl mkl-devel mkl_fft numexpr mingw make setuptools nose scikit-image six scipy dxchange
conda update --all
```

Neither machine has a  `usr` folder underneath `Library`.  There is a `Anaconda3\Library\bin` folder, but there is not an `install.exe` there.  

So, I don't know why your system is different.   Perhaps `install.exe` included in Miniconda but not Anaconda?  Is there a package that needs to be installed that is not listed in the appveyor config?  
The output of `conda list`  on one of my machines is given below.


You said:

> > But install is not made available by the conda environment defined for tomopy. ...

> I don't believe this is true

What package provides "install.exe"? 



conda list results:

```
# packages in environment at C:\Users\newville.CARS_NT_DOMAIN\AppData\Local\Continuum\anaconda3:
#
# Name                    Version                   Build  Channel
_ipyw_jlab_nb_ext_conf    0.1.0                    py36_0  
alabaster                 0.7.12                   py36_0  
anaconda                  custom           py36h363777c_0  
anaconda-client           1.7.2                    py36_0  
anaconda-navigator        1.9.2                    py36_0  
anaconda-project          0.8.2                    py36_0  
appdirs                   1.4.3            py36h28b3542_0  
asn1crypto                0.24.0                   py36_0  
asteval                   0.9.13           py36h39e3cac_0    gsecars
astroid                   2.1.0                    py36_0  
astropy                   3.0.5            py36he774522_0  
atomicwrites              1.2.1                    py36_0  
attrs                     18.2.0           py36h28b3542_0  
automat                   0.7.0                    py36_0  
babel                     2.6.0                    py36_0  
backcall                  0.1.0                    py36_0  
backports                 1.0                      py36_1  
backports.os              0.1.1                    py36_0  
backports.shutil_get_terminal_size 1.0.0                    py36_2  
beautifulsoup4            4.6.3                    py36_0  
bitarray                  0.8.3            py36hfa6e2cd_0  
bkcharts                  0.2              py36h7e685f7_0  
blas                      1.0                         mkl  
blaze                     0.11.3                   py36_0  
bleach                    3.0.2                    py36_0  
blosc                     1.14.4               he51fdeb_0  
bokeh                     1.0.2                    py36_0  
boto                      2.49.0                   py36_0  
bottleneck                1.2.1            py36h452e1ab_1  
bzip2                     1.0.6                hfa6e2cd_5  
ca-certificates           2018.03.07                    0  
certifi                   2018.10.15               py36_0  
cffi                      1.11.5           py36h74b6da3_1  
cftime                    1.0.2.1          py36h8c2d366_0  
chardet                   3.0.4                    py36_1  
click                     7.0                      py36_0  
cloudpickle               0.6.1                    py36_0  
clyent                    1.2.2                    py36_1  
colorama                  0.4.0                    py36_0  
comtypes                  1.1.7                    py36_0  
conda                     4.5.11                   py36_0  
conda-build               3.17.0                   py36_0  
conda-env                 2.6.0                         1  
conda-verify              3.1.1                    py36_0  
console_shortcut          0.1.1                         3  
constantly                15.1.0           py36h28b3542_0  
constructor               2.3.0                    py36_0  
contextlib2               0.5.5            py36he5d52c0_0  
cryptography              2.3.1            py36h74b6da3_0  
curl                      7.62.0               h2a8f88b_0  
cycler                    0.10.0           py36h009560c_0  
cython                    0.29             py36ha925a31_0  
cytoolz                   0.9.0.1          py36hfa6e2cd_1  
dask                      0.19.1                   py36_0  
dask-core                 1.0.0                    py36_0  
datashape                 0.5.4                    py36_1  
decorator                 4.3.0                    py36_0  
distributed               1.25.0                   py36_0  
docutils                  0.14             py36h6012d8f_0  
dxchange                  0.1.4                    py36_0    gsecars
dxfile                    0.4                      py36_0    gsecars
edffile                   5.0.0                    py36_0    gsecars
entrypoints               0.2.3                    py36_2  
et_xmlfile                1.0.1            py36h3d2d736_0  
fastcache                 1.0.2            py36hfa6e2cd_2  
filelock                  3.0.10                   py36_0  
flask                     1.0.2                    py36_1  
flask-cors                3.0.7                    py36_0  
freetype                  2.9.1                ha9979f8_1  
future                    0.17.1                   py36_0  
get_terminal_size         1.0.0                h38e98db_0  
gevent                    1.3.7            py36he774522_1  
glob2                     0.6                      py36_1  
greenlet                  0.4.15           py36hfa6e2cd_0  
h5py                      2.8.0            py36h3bdd7fb_2  
hdf4                      4.2.13               h712560f_2  
hdf5                      1.10.2               hac2f561_1  
heapdict                  1.0.0                    py36_2  
html5lib                  1.0.1                    py36_0  
hyperlink                 18.0.0                   py36_0  
icc_rt                    2017.0.4             h97af966_0  
icu                       58.2                 ha66f8fd_1  
idna                      2.7                      py36_0  
imageio                   2.4.1                    py36_0  
imagesize                 1.1.0                    py36_0  
importlib_metadata        0.6                      py36_0  
incremental               17.5.0                   py36_0  
intel-openmp              2019.1                      144  
ipykernel                 5.1.0            py36h39e3cac_0  
ipython                   7.2.0            py36h39e3cac_0  
ipython_genutils          0.2.0            py36h3c5d0ee_0  
ipywidgets                7.4.2                    py36_0  
isort                     4.3.4                    py36_0  
itsdangerous              1.1.0                    py36_0  
jdcal                     1.4                      py36_0  
jedi                      0.13.1                   py36_0  
jinja2                    2.10                     py36_0  
jpeg                      9b                   hb83a4c4_2  
jsonschema                2.6.0            py36h7636477_0  
jupyter                   1.0.0                    py36_7  
jupyter_client            5.2.3                    py36_0  
jupyter_console           6.0.0                    py36_0  
jupyter_core              4.4.0                    py36_0  
jupyterlab                0.35.3                   py36_0  
jupyterlab_launcher       0.13.1                   py36_0  
jupyterlab_server         0.2.0                    py36_0  
keyring                   16.1.1                   py36_0  
kiwisolver                1.0.1            py36h6538335_0  
krb5                      1.16.1               h038dc86_6  
lazy-object-proxy         1.3.1            py36hfa6e2cd_2  
libarchive                3.3.3                h798a506_1  
libcurl                   7.62.0               h2a8f88b_0  
libiconv                  1.15                 h1df5818_7  
liblief                   0.9.0                ha925a31_0  
libnetcdf                 4.6.1                h4eace20_2  
libpng                    1.6.35               h2a8f88b_0  
libpq                     10.5                 h5fe2233_0  
libsodium                 1.0.16               h9d3ae62_0  
libssh2                   1.8.0                hd619d38_4  
libtiff                   4.0.9                h36446d0_2  
libxml2                   2.9.8                hadb2253_1  
libxslt                   1.1.32               hf6f1972_0  
llvmlite                  0.26.0           py36ha925a31_0  
lmfit                     0.9.12                   py36_0    gsecars
locket                    0.2.0            py36hfed976d_1  
lxml                      4.2.5            py36hef2cd61_0  
lz4-c                     1.8.1.2              h2fa13f4_0  
lzo                       2.10                 h6df0209_2  
m2w64-gcc-libgfortran     5.3.0                         6  
m2w64-gcc-libs            5.3.0                         7  
m2w64-gcc-libs-core       5.3.0                         7  
m2w64-gmp                 6.1.0                         2  
m2w64-libwinpthread-git   5.0.0.4634.697f757               2  
make                      3.82                          0  
markupsafe                1.1.0            py36he774522_0  
matplotlib                3.0.1            py36hc8f65d3_0  
mccabe                    0.6.1                    py36_1  
menuinst                  1.4.14           py36hfa6e2cd_0  
mingw                     4.7                           1  
mistune                   0.8.4            py36he774522_0  
mkl                       2018.0.3                      1  
mkl-devel                 2018.0.3                      1  
mkl-include               2018.0.3                      1  
mkl-service               1.1.2            py36hb217b18_5  
mkl_fft                   1.0.6            py36hdbbee80_0  
mkl_random                1.0.1            py36h77b88f5_1  
more-itertools            4.3.0                    py36_0  
mpmath                    1.0.0                    py36_2  
msgpack-python            0.5.6            py36he980bc4_1  
msys2-conda-epoch         20160418                      1  
multipledispatch          0.6.0                    py36_0  
navigator-updater         0.2.1                    py36_0  
nbconvert                 5.3.1                    py36_0  
nbformat                  4.4.0            py36h3a5bc1b_0  
netcdf4                   1.4.2            py36h9ba30bf_0  
networkx                  2.2                      py36_1  
nltk                      3.3.0                    py36_0  
nose                      1.3.7                    py36_2  
notebook                  5.7.2                    py36_0  
nsis                      3.01                          8  
numba                     0.39.0           py36h830ac7b_0  
numexpr                   2.6.8            py36h9ef55f4_0  
numpy                     1.15.4           py36ha559c80_0  
numpy-base                1.15.4           py36h8128ebf_0  
numpydoc                  0.8.0                    py36_0  
odo                       0.5.1            py36h7560279_0  
olefile                   0.46                     py36_0  
openpyxl                  2.5.11                   py36_0  
openssl                   1.0.2p               hfa6e2cd_0  
packaging                 18.0                     py36_0  
pandas                    0.23.4           py36h830ac7b_0  
pandoc                    2.2.3.2                       0  
pandocfilters             1.4.2                    py36_1  
parso                     0.3.1                    py36_0  
partd                     0.3.9                    py36_0  
path.py                   11.5.0                   py36_0  
pathlib2                  2.3.2                    py36_0  
patsy                     0.5.1                    py36_0  
pcre                      8.42                 hd6b2f15_0  
peakutils                 1.1.0                    py36_0    gsecars
pep8                      1.7.1                    py36_0  
pickleshare               0.7.5                    py36_0  
pillow                    5.3.0            py36hdc69c19_0  
pip                       18.1                     py36_0  
pkginfo                   1.4.2                    py36_1  
pluggy                    0.8.0                    py36_0  
ply                       3.11                     py36_0  
prometheus_client         0.4.2                    py36_0  
prompt_toolkit            2.0.7                    py36_0  
psutil                    5.4.8            py36he774522_0  
psycopg2                  2.7.5            py36h74b6da3_0  
py                        1.7.0                    py36_0  
py-lief                   0.9.0            py36ha925a31_0  
pyasn1                    0.4.4            py36h28b3542_0  
pyasn1-modules            0.2.2                    py36_0  
pycifrw                   4.3                      py36_0    gsecars
pycodestyle               2.4.0                    py36_0  
pycosat                   0.6.3            py36hfa6e2cd_0  
pycparser                 2.19                     py36_0  
pycrypto                  2.6.1            py36hfa6e2cd_9  
pycurl                    7.43.0.2         py36h74b6da3_0  
pyepics                   3.3.0            py36hcbe5ab8_0    gsecars
pyflakes                  2.0.0                    py36_0  
pygments                  2.2.0            py36hb010967_0  
pyhamcrest                1.9.0                    py36_2  
pylint                    2.2.2                    py36_0  
pyodbc                    4.0.24           py36h6538335_0  
pyopenssl                 18.0.0                   py36_0  
pyparsing                 2.3.0                    py36_0  
pyqt                      5.9.2            py36h6538335_2  
pyshortcuts               1.0                       <pip>
pyshortcuts               1.3                      py36_0    gsecars
pysocks                   1.6.8                    py36_0  
pytables                  3.4.4            py36he6f6034_0  
pytest                    4.0.1                    py36_0  
pytest-arraydiff          0.2              py36h39e3cac_0  
pytest-astropy            0.4.0                    py36_0  
pytest-doctestplus        0.2.0                    py36_0  
pytest-openfiles          0.3.1                    py36_0  
pytest-remotedata         0.3.1                    py36_0  
python                    3.6.7                h33f27b4_1  
python-dateutil           2.7.5                    py36_0  
python-libarchive-c       2.8                      py36_6  
pytz                      2018.7                   py36_0  
pywavelets                1.0.1            py36h8c2d366_0  
pywin32                   223              py36hfa6e2cd_1  
pywinpty                  0.5.4                    py36_0  
pyyaml                    3.13             py36hfa6e2cd_0  
pyzmq                     17.1.2           py36hfa6e2cd_0  
qt                        5.9.6            vc14h1e9a669_2  
qtawesome                 0.5.3                    py36_0  
qtconsole                 4.4.2                    py36_0  
qtpy                      1.5.2                    py36_0  
requests                  2.20.1                   py36_0  
rope                      0.11.0                   py36_0  
ruamel_yaml               0.15.46          py36hfa6e2cd_0  
scikit-image              0.14.0           py36h6538335_1  
scikit-learn              0.20.1           py36hb854c30_0  
scipy                     1.1.0            py36h4f6bf74_1  
seaborn                   0.9.0                    py36_0  
send2trash                1.5.0                    py36_0  
service_identity          17.0.0           py36h28b3542_0  
setuptools                40.6.2                   py36_0  
simplegeneric             0.8.1                    py36_2  
singledispatch            3.4.0.3          py36h17d0c80_0  
sip                       4.19.8           py36h6538335_0  
six                       1.11.0                   py36_1  
snappy                    1.1.7                h777316e_3  
snowballstemmer           1.2.1            py36h763602f_0  
sortedcollections         1.0.1                    py36_0  
sortedcontainers          2.0.5                    py36_0  
spefile                   1.6                      py36_0    gsecars
sphinx                    1.8.2                    py36_0  
sphinxcontrib             1.0                      py36_1  
sphinxcontrib-websupport  1.1.0                    py36_1  
spyder                    3.3.1                    py36_1  
spyder-kernels            1.0.1                    py36_0  
sqlalchemy                1.2.14           py36he774522_0  
sqlite                    3.25.3               he774522_0  
statsmodels               0.9.0            py36h452e1ab_0  
sympy                     1.3                      py36_0  
tblib                     1.3.2            py36h30f5020_0  
termcolor                 1.1.0                    py36_1  
terminado                 0.8.1                    py36_1  
testpath                  0.4.2                    py36_0  
tifffile                  0.14.0                   py36_1    gsecars
tk                        8.6.8                hfa6e2cd_0  
tomopy                    1.1.4            py36h39e3cac_0    gsecars
tomopy                    1.1.4                     <pip>
toolz                     0.9.0                    py36_0  
tornado                   5.1.1            py36hfa6e2cd_0  
tqdm                      4.28.1           py36h28b3542_0  
traitlets                 4.3.2            py36h096827d_0  
twisted                   18.9.0           py36he774522_0  
typed-ast                 1.1.0            py36hfa6e2cd_0  
typing                    3.6.4                    py36_0  
uncertainties             3.0.2                    py36_0    gsecars
unicodecsv                0.14.1           py36h6450c06_0  
urllib3                   1.23                     py36_0  
vc                        14.1                 h0510ff6_4  
vs2015_runtime            14.15.26706          h3a45250_0  
wcwidth                   0.1.7            py36h3d5aa90_0  
webencodings              0.5.1                    py36_1  
werkzeug                  0.14.1                   py36_0  
wheel                     0.32.3                   py36_0  
widgetsnbextension        3.4.2                    py36_0  
win_inet_pton             1.0.1                    py36_1  
win_unicode_console       0.5              py36hcdbd4b5_0  
wincertstore              0.2              py36h7fe50ca_0  
winpty                    0.4.3                         4  
wrapt                     1.10.11          py36hfa6e2cd_2  
wxmplot                   0.9.32           py36h39e3cac_0    gsecars
wxpython                  4.0.3            py36h6538335_0  
wxutils                   0.2                      py36_0    gsecars
xlrd                      1.1.0                    py36_1  
xlsxwriter                1.1.2                    py36_0  
xlwings                   0.14.1                   py36_0  
xlwt                      1.3.0            py36h1a4751e_0  
xraylarch                 0.9.38                    <pip>
xraylarch                 0.9.40                    <pip>
xraylarch                 0.9.40           py36h39e3cac_0    gsecars
xz                        5.2.4                h2fa13f4_4  
yaml                      0.1.7                hc54c509_2  
zeromq                    4.2.5                he025d50_1  
zict                      0.1.3                    py36_0  
zlib                      1.2.11               h62dcd97_3  
zope                      1.0                      py36_1  
zope.interface            4.6.0            py36he774522_0  
zstd                      1.3.3                hfe6a214_0  
```

**Comment by carterbox** [comment]
@newville I compared our package lists and didn't see anything that looked significantly different. I am going to try a clean install of latest Miniconda and Anaconda (because they only offer the latest installers). I will report back when I figure out how `install.exe` was added to my systems.

**Comment by prjemian** [comment]
FWIW, on my Win10-64 system, `install.exe` shows up in Anaconda in two places:

```
C:\Users\Pete\Apps\Anaconda\pkgs\git-2.19.2-1000\Library\usr\bin\install.exe
C:\Users\Pete\Apps\Anaconda\envs\bluesky\Library\usr\bin\install.exe
```

**Comment by newville** [comment]
@prjemian  Thanks.  That implies that the `git` package would be required.  

Personally, I think the better approach would be to not use `install` or `make install` (or `make clean` for that matter). in `build.py` at all.  That is, the `make install` just copies the shared library from one location (`tomopy/sharedlibs`) to another (`/tomopy/build/lib.PLAT-VERS/tomopy/sharedlibs/`).  Since the shared lib will not be in use and copy permission will not be an issue, `install` could be replaced with built-in `shutil.copy` and `os.chmod`.   That needs no external packages at all.  Similarly, `make clean` could use `glob` and `os.unlink` instead of spawning an external process (`make`) to launch a shell to run a program named `rm`.  

I've pushed changes to do this to my own fork of tomopy and verified that building, installing, and cleaning work for me on Windows, MacOS, and Linux.  I could make a PR for this, but whether this is desirable for the main repo has become somewhat unclear to me.

**Comment by dgursoy** [comment]
> If you intend to drop Windows support for tomopy in favor of docker or some other container system, please let me know. We will fork and continue to use and support tomopy as a library for these and other downstream applications that might want to work on Windows as well as MacOS or Linux.

@newville There is no reason to be worried. We don't have any intention of dropping support for Windows users. These options were to discuss to find the best ways of providing this support with minimal developer effort, and honestly I find such conversations quite informative, but also it doesn't necessarily mean that the packaging will be changed.  

> Since getting tomopy building on Windows earlier this fall, our group has committed to using tomopy for our Xray Microprobe Map processing GUI to be able to process and view first-generation fluorescence and diffraction tomography data, and a visiting student has written a very nice GUI for on top of tomopy for processing and viewing the full-field tomography data from APS 13BM.

@newville I know there are many users who would enjoy a GUI for TomoPy. Is there any chance that we can place the GUI in the TomoPy repository? If the dependencies are lightweight, we may also consider distributing it with TomoPy. In any case we should provide the install links in the TomoPy docs.

**Comment by newville** [comment]
On Wed, Dec 5, 2018 at 7:24 PM Jonathan R. Madsen <notifications@github.com>
wrote:

> @newville <https://github.com/newville>:
>
> What package provides "install.exe"?
>
> conda install -c msys2 m2-coreutils or conda install -c msys2 posix
>

Yes, but these packages are *not* listed as required to build tomopy
(either requirements.txt or .appveyor.yml).  Hence the issue.
Adding one of these as a dependency would be possible solution.

I think it would be better to use builtin python functionality
(`shutil.copy`, 'os.chmod', 'os.unlink') rather than installing extra
dependencies to have  `make` run external shell commands.

Tomopy is a library.
>
> Agree.
>
> With its clean separation of C library and python interface, supporting it
> on multiple platforms should not be hard.
>
> Disagree. With years of experience to back it up.
>

Well, currently all the tomopy extensions are in plain C, and are not too
hard to support.  I'm very glad to read that you have years of experience,
and it is always good to have confidence in ones abilities.

I am working on extensions that involve MPI and CUDA and supporting these
> will essentially require eliminating the current build system and replacing
> it with CMake (unless you want even more build system issues than what
> currently exist). Also, tomopy will start using C++11 in these extensions
> and a C++ compiler is much, much more complex than a C compiler and MinGW
> is, again, a *non-native* compiler to Windows. So if issues arise on
> Windows, I will only be able to do so much because NERSC does not pay me to
> support Windows.
>

Hm, this sounds quite a bit like you asserting that you will be changing
tomopy, and that supporting Windows is a very low priority when you do make
such changes. I would assume that you have worked out a plan with tomopy
developers for what you plan to do, but what you say here seems at odds
with what I have heard from tomopy developers.  That is sort of puzzling.

I don't see why using tomopy should require a docker container.
>
> The idea wasn't proposing that tomopy itself would require a docker
> container. The very nature of being able to put tomopy into a container
> would mean it was supported by Linux and by extension macOS because tomopy
> doesn't do anything in the space that separates Linux from BSD. Thus, it
> would only be a requirement for Windows users.
>

You are definitely saying that tomopy would require a docker container to
use on Windows.  I can imagine that some functionality would not work on
all systems, for example that some HPC code might only run on posix
systems.  I do not see why existing code and functionality should be made
to stop working.

... visiting student has written a very nice GUI for on top of tomopy
>
> Is the GUI directly linking to libpython? Directly linking to libtomopy?
> Because if that's not the case, there isn't an issue here.
>

Both of the GUIs are written in python, import tomopy and use tomopy
functions to do calculations on data and display results.  The one from the
student allows the sort-of-experienced tomography user to select different
methods and parameters for centering, padding, filtering, etc and visualize
the results, and to then do a set of reconstructions.  The other is a more
specific task to visualizing fluorescence tomography data, within a larger
application, but this also allows the user to adjust filtering, centering,
etc.  These apps are meant to run as interactive programs for visualizing
data as it is being collected, and to be readily installable to run on
users desktops and laptops.

They do rely on Python dynamically loading the libtomopy shared library.
Is that what you mean by "direct linking"?  That's not how I would describe
how Python works with shared libraries, but these apps do rely on loading
libtomopy.

We cannot distribute these to the majority of our users without having
> tomopy available on Windows
>
> Please elaborate how these GUIs are using tomopy. There is nothing I can
> think of that would prohibit these GUIs from running tomopy with Docker +
> Windows unless the GUI is directly linking to Python or the tomopy
> libraries.
>

They are written in Python and import and use tomopy.


> If you are just executing processes under the hood of the GUI then python
> ./run-some-tomopy-script.py would just be docker run -it tomopy:latest
> ./run-some-tomopy-script.py.
>

That is not how these GUIs work.

In fact, w.r.t. distribution, using docker on your end would make
> distributing your GUI *more* portable because then the GUI:
>
>    - could be run on any Windows, macOS, or Windows system
>    - require zero configuration (since everything built and bundled in
>    docker container) aside from installing up an Xming server on Windows and
>    XQuartz on macOS
>
>
We are not going to do any of that.


>
> We will fork and continue to use and support tomopy as a library for these
> and other downstream applications that might want to work on Windows as
> well as MacOS or Linux
>
> That really doesn't seem necessary and it is possible that doing this
> would add more work on your end over time because this option requires
> having to maintain a separate code base continuously vs. doing a one-time
> patch to the GUIs to interact with tomopy inside a docker container.
>
And the point of using docker is that the containers can run on Windows as
> well as macOS or Linux but decreases the burden on the developers to deal
> with explicit cross-platform compatibility issues which, again, are not
> always trivial.
>

I can see how a docker container could make your life easier, but we're not
going have our applications use tomopy from within a docker container.

**Comment by newville** [comment]
Hi Doga,

On Thu, Dec 6, 2018 at 10:52 AM Doga Gursoy <notifications@github.com>
wrote:

> If you intend to drop Windows support for tomopy in favor of docker or
> some other container system, please let me know. We will fork and continue
> to use and support tomopy as a library for these and other downstream
> applications that might want to work on Windows as well as MacOS or Linux.
>
> @newville <https://github.com/newville> There is no reason to be worried.
> We don't have any intention of dropping support for Windows users. These
> options were to discuss to find the best ways of providing this support
> with minimal developer effort, and honestly I find such conversations quite
> informative, but also it doesn't necessarily mean that the packaging will
> be changed.
>
OK.

> Since getting tomopy building on Windows earlier this fall, our group has
> committed to using tomopy for our Xray Microprobe Map processing GUI to be
> able to process and view first-generation fluorescence and diffraction
> tomography data, and a visiting student has written a very nice GUI for on
> top of tomopy for processing and viewing the full-field tomography data
> from APS 13BM.
>
> @newville <https://github.com/newville> I know there are many users who
> would enjoy a GUI for TomoPy. Is there any chance that we can place the GUI
> in the TomoPy repository? If the dependencies are lightweight, we may also
> consider distributing it with TomoPy. In any case we should provide the
> install links in the TomoPy docs.
>

The GUI code from Brandt Gibson is at
https://github.com/bramgibs/TomoPy_GUI.  There is not a lot of
documentation, but the code is actually pretty small, and fairly clean in
its approach.  Again, this is from a visiting PhD student in Geology who
had very little Python or tomopy experience before starting to write this
GUI this fall.

I believe this can only read APS13BM data, but it might not take much
effort to make it more general.  If you have someone interested in working
on or extending this, I'd be happy to help as much as I can.

I have not done this yet, but it is my intention to wrap this with a conda
constructor file to make a single-file installer.

**Comment by dgursoy** [comment]
@newville Great! We worked out some parts of this project with Brandt during summer, so I know some background. I'd be happy to add other readers for tomography on a per request basis. I assume we can start with 2-BM and 32-ID (@decarlof). We should then add some documentation in TomoPy for other users to find the repo.

**Comment by jrmadsen** [comment]
@newville 

> Yes, but these packages are *not* listed as required to build tomopy
(either requirements.txt or .appveyor.yml).  

As @carterbox and I noted, this problem had not shown up in any of our personal testing on Windows nor in AppVeyor. 

> They are written in Python and import and use tomopy.

This means my concerns about linking do not apply and that there would be no issues with Windows + Docker + tomopy + your GUIs but your later statements suggest that these discussions will not be productive so I am dropping it. 

As I stated previously, I will do my best to make sure any modifications don't break Windows builds but can only spend so much time doing so.

**Comment by carterbox** [comment]
I merged a patch today to address the original issue. Good discussion, folks.
